File: tools/shv-interface-codegen/tests/owner_test.cpp
#include "shv_interface.hpp"
#include "test_support.hpp"
using namespace symphony::authoring;
struct InterfaceTests {
  Arguments args;
  Json r, d;
  explicit InterfaceTests(const Arguments &selected) : args(selected) {
    require(!args.registration.empty(), "--registration is required");
    const auto loaded = shv::load(args.root, args.registration);
    r = loaded.first;
    d = loaded.second;
  }
  void test_frozen_descriptor() {
    const auto hist =
        read_json(args.root / string(r.at("history"))).at("descriptors");
    test::descriptor(args.engine, shv::latest(hist), d.at("current_version"));
  }
  void test_generated_parity() {
    emit(args.root, shv::owner_outputs(r, d), true, true);
    test::check(shv::owner_outputs(r, d) == shv::owner_outputs(r, Json(d)),
                "deterministic projections");
  }
  void test_future_owner_and_collision() {
    {
      TempDir temp;
      auto future = d;
      future["module_id"] = "user-sensor-engine";
      future["engine_id"] = "user-sensor";
      future["namespace"] = "user::sensor";
      future["current_version"] = "1.0.0";
      future["releases"] = {{"1.0.0", shv::names(d.at("operations"))}};
      for (const auto *key : {"schemas", "companions", "embedded_dependencies"})
        future[key] = Json::array();
      write_json(temp.path / "declaration.json", future);
      write_bytes(temp.path / "history.json", "{\"descriptors\":{}}");
      auto registration = r;
      registration["declaration"] = "declaration.json";
      registration["history"] = "history.json";
      registration["history_digest"] =
          digest_bytes(read_bytes(temp.path / "history.json"));
      registration["go_title"] = "UserSensor";
      registration["go_stem"] = "user_sensor";
      registration["cpp_output"] = "out/sensor.hpp";
      registration["go_output"] = "out/sensor.go";
      registration["cmake_output"] = "out/sensor.cmake";
      registration["embedded_compatibility_version"] = nullptr;
      registration["embedded_macro"] = "USER_SENSOR_PREVIOUS";
      write_json(temp.path / "registration.json", registration);
      const auto [reg, decl] =
          shv::load(temp.path, temp.path / "registration.json");
      const auto outputs = shv::owner_outputs(reg, decl);
      emit(temp.path, outputs, false, true);
      emit(temp.path, outputs, true, true);
      auto other = decl;
      other["module_id"] = "other-sensor-engine";
      other["engine_id"] = "other-sensor";
      test::rejects(
          [&] { emit(temp.path, shv::owner_outputs(reg, other), false, true); },
          "other owner collision");
      for (const auto &[path, body] : outputs)
        test::check(read_bytes(temp.path / path) == body,
                    "preflight leaves prior output unchanged");
      write_bytes(temp.path / "out/sensor.go", "caller-owned");
      test::rejects([&] { emit(temp.path, outputs, false, true); },
                    "caller output collision");
      test::check(read_bytes(temp.path / "out/sensor.go") == "caller-owned",
                  "caller bytes unchanged");
    }
  }
  void test_registration_rejections() {
    const std::vector<std::pair<std::string, std::function<void(Json &)>>>
        mutations{
            {"path", [](Json &x) { x["cpp_output"] = "../escape"; }},
            {"symbol", [](Json &x) { x["go_title"] = "bad;code"; }},
            {"collision", [](Json &x) { x["cpp_output"] = x["declaration"]; }},
            {"history",
             [](Json &x) {
               x["history_digest"] = "sha256:" + std::string(64, '0');
             }},
            {"embedded",
             [](Json &x) { x["embedded_compatibility_version"] = "latest"; }}};
    for (const auto &[name, mutate] : mutations) {
      TempDir temp;
      auto bad = r;
      mutate(bad);
      write_json(temp.path / "reg.json", bad);
      test::rejects([&] { shv::load(args.root, temp.path / "reg.json"); },
                    name);
    }
  }
};
int main(int argc, char **argv) {
  return main_guard([&] {
    const Arguments args(argc, argv);
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_frozen_descriptor\n";
      suite.test_frozen_descriptor();
      std::cout << "PASS test_frozen_descriptor\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_generated_parity\n";
      suite.test_generated_parity();
      std::cout << "PASS test_generated_parity\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_future_owner_and_collision\n";
      suite.test_future_owner_and_collision();
      std::cout << "PASS test_future_owner_and_collision\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_registration_rejections\n";
      suite.test_registration_rejections();
      std::cout << "PASS test_registration_rejections\n";
    }
    test::done();
  });
}

File: tools/shv-interface-codegen/tests/owner_test.py
import argparse, copy, hashlib, importlib.util, json, subprocess, tempfile, unittest
from pathlib import Path

H = Path(__file__).resolve().parents[1]
R = H.parents[1]
s = importlib.util.spec_from_file_location("owner_codegen", H / "owner_codegen.py")
g = importlib.util.module_from_spec(s)
s.loader.exec_module(g)
p = argparse.ArgumentParser()
p.add_argument("--registration", type=Path, required=True)
p.add_argument("--engine", required=True)
a, rest = p.parse_known_args()


class OwnerTests(unittest.TestCase):
    def test_frozen_descriptor(self):
        r, d = g.load(R, a.registration)
        hist = g.read(R / r["history"])["descriptors"]
        expected = copy.deepcopy(list(hist.values())[-1])
        expected["engine_version"] = d["current_version"]
        expected.pop("descriptor_digest")
        expected["descriptor_digest"] = g.digest(expected)
        self.assertEqual(
            json.loads(subprocess.check_output([a.engine, "--descriptor"])), expected
        )

    def test_generated_parity(self):
        r, d = g.load(R, a.registration)
        g.emit(R, r, d, True)
        self.assertEqual(g.outputs(r, d), g.outputs(r, copy.deepcopy(d)))

    def test_future_owner_and_collision(self):
        r, d = g.load(R, a.registration)
        with tempfile.TemporaryDirectory() as td:
            root = Path(td)
            d = copy.deepcopy(d)
            d.update(
                module_id="user-sensor-engine",
                engine_id="user-sensor",
                namespace="user::sensor",
                current_version="1.0.0",
                releases={"1.0.0": [o["operation_name"] for o in d["operations"]]},
                schemas=[],
                companions=[],
                embedded_dependencies=[],
            )
            (root / "declaration.json").write_text(json.dumps(d))
            (root / "history.json").write_text('{"descriptors":{}}')
            r.update(
                declaration="declaration.json",
                history="history.json",
                history_digest="sha256:"
                + hashlib.sha256((root / "history.json").read_bytes()).hexdigest(),
                go_title="UserSensor",
                go_stem="user_sensor",
                cpp_output="out/sensor.hpp",
                go_output="out/sensor.go",
                cmake_output="out/sensor.cmake",
                embedded_compatibility_version=None,
                embedded_macro="USER_SENSOR_PREVIOUS",
            )
            (root / "registration.json").write_text(json.dumps(r))
            reg, decl = g.load(root, root / "registration.json")
            g.emit(root, reg, decl)
            g.emit(root, reg, decl, True)
            other = copy.deepcopy(decl)
            other.update(module_id="other-sensor-engine", engine_id="other-sensor")
            before = {path: (root / path).read_bytes() for path in g.outputs(reg, decl)}
            with self.assertRaises(ValueError):
                g.emit(root, reg, other)
            self.assertEqual(
                before, {path: (root / path).read_bytes() for path in before}
            )
            (root / "out/sensor.go").write_text("caller-owned")
            with self.assertRaises(ValueError):
                g.emit(root, reg, decl)
            self.assertEqual((root / "out/sensor.go").read_text(), "caller-owned")

    def test_registration_rejections(self):
        r, d = g.load(R, a.registration)
        for name, mutate in [
            ("path", lambda x: x.update(cpp_output="../escape")),
            ("symbol", lambda x: x.update(go_title="bad;code")),
            ("collision", lambda x: x.update(cpp_output=x["declaration"])),
            ("history", lambda x: x.update(history_digest="sha256:" + "0" * 64)),
            ("embedded", lambda x: x.update(embedded_compatibility_version="latest")),
        ]:
            with self.subTest(name=name), tempfile.TemporaryDirectory() as td:
                bad = copy.deepcopy(r)
                mutate(bad)
                p = Path(td) / "reg.json"
                p.write_text(json.dumps(bad))
                with self.assertRaises(ValueError):
                    g.load(R, p)


if __name__ == "__main__":
    unittest.main(argv=["owner-test", *rest])
